Refactor TDB internal bounds computation

kyle-cypress opened the pull request:
### Summary of changes
Refactor internal TDB flash bounds computation to reduce code duplication and to provide more viable default values on parts where the program and erase block sizes are close or identical in value.

The default TDBStore internal flash size is determined as the larger of two flash sectors or 14 flash pages (it previously defaulted to two sectors). This increases the usability of the default size computation on devices where the flash sector and page sizes are close or equal to each other.

#### Impact of changes
Applications which use TDBStore (or another class which wraps it), and do not explicitly specify a value for `internal_size` in `storage_tdb_internal`, will consume additional flash space for TDBStore. 

#### Migration actions required
If an application does not already specify `internal_size` for `storage_tdb_internal` in its mbed_app.json, and the increased storage is not desired, add an entry to mbed_app.json to explicitly specify the desired size.
If an application already specifies `internal_size` for `storage_tdb_internal` in its mbed_app.json, no changes are required.

### Documentation
N/A

----------------------------------------------------------------------------------------------------------------
### Pull request type (*required*)
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results (*required*)

    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
----------------------------------------------------------------------------------------------------------------
### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

Commit:
Refactor internal flash TDB bounds determination

Default the size to the larger of two sectors or 10 pages, so that the
  computation works better on devices with a low sector to page size ratio.
Reduce code duplication.

Commit:
Move TDB bounds computation for better reuse

Migrate into TDBStore so that DirectAccessDeviceKey can use it as well.

VeijoPesonen commented on features/storage/kvstore/tdbstore/TDBStore.cpp:
```diff
@@ -1447,6 +1453,136 @@ int TDBStore::reserved_data_get(void *reserved_data, size_t reserved_data_buf_si
+        ret = MBED_ERROR_MEDIA_FULL;
+    } else {
+        *start_address = curr_addr;
+    }
```
@SeppoTakalo should we maybe use MBED_ERROR here? To replace all the highlighted lines.

SeppoTakalo replied:
Would be too big API change.. maybe for the OS version 6.

VeijoPesonen commented on features/storage/kvstore/conf/kv_config.cpp:
```diff
@@ -689,57 +548,71 @@ MBED_WEAK BlockDevice *get_other_blockdevice()
 
-    //Check that internal flash has 2 or more sectors
-    if (_calculate_blocksize_match_tdbstore(kvstore_config.internal_bd) != MBED_SUCCESS) {
+    //Check if TDBStore has at least 2 sector.
```
*s or 10 pages.

JanneKiiskila replied:
Why would 10 pages work instead of a full sector? I think TDBSTore uses the flash sectors to do the garbage collection, in sort of A vs. B, one of them being active and the other one as the spare?

kyle-cypress replied:
Yes, that matches my understanding of TDBStore's garbage collection - it divides the available flash area in half and alternates between the two to do copying GC.
The problem is that the default approach breaks down when the sector size and page sizes are the same (or close), because many of the KVStore operations consume an entire page. Based on our analysis of the KVStore greentea tests as a proxy for typical usage, the minimum page count appeared to be around 10 i.e. 5 pages in active use, 5 in reserve for GC (or 2 sectors if that is larger, so that GC continues to work). If there's a different number of pages that more directly corresponds to something intrinsic in the driver I'd be happy to go with that instead.

kyle-cypress replied:
@VeijoPesonen, with regard to:
>*s or 10 pages.

Wouldn't expanding this check produce a backwards compatibility concern? Or does this qualify as a situation where it would probably have always failed, just in a less explicit manner, and therefore it is okay to add a new error condition?

kyle-cypress replied:
@VeijoPesonen, can you please confirm that the expanded check is not a backwards compatibility concern?

geky replied:
Hi, was browsing some PRs randomly, sorry for jumping in.

The number of pages TDBStore needs depends on the number of KV pairs+1. So for 4 KV pairs it needs 5 pages. (Actually it may need 2 pages per KV pair, not sure).

Unfortunately I think this means you can't make this check unless you know the number of KV pairs beforehand.

kyle-cypress replied:
I agree that you can't compute a minimum that is guaranteed to be enough unless you know how much is intended to be stored. However, there is a bare minimum size (e.g. space for internal bookkeeping, a second sector for garbage collection) below which TDBStore cannot function. That is what this check is intended to achieve. Are you saying that the bare minimum page count calculation should be different?

geky replied:
Ah, my understanding was that the theoretical minimum would be 4 pages total (2 for the active master record, 2 for the old master record, where pages = program size'd units). But my understand could be out of date.

Out of curiousity do you know where the other 2x3 pages come from?

kyle-cypress replied:
I not able to determine precisely where that came from.
However, after further consideration I've removed the hard-fail on too few pages given that the exact minimum is application specific. Full reasoning is in the commit message for a3a4eef59939741999101f1eb5b4367f422f9e6e

VeijoPesonen commented on features/storage/kvstore/tdbstore/TDBStore.h:
```diff
@@ -280,6 +280,33 @@ class TDBStore : public KVStore {
+
+    /**
+     * @brief Get the TDBStore flash bounds from the configured start address and size.
+     * Configured start address/size must not both be 0.
```
Why? If size zero indicates that blockdevice can take over the whole flash region what makes difference between address zero and any other address? Flash can start from address 0x0.

0xc0170 replied:
@kyle-cypress ?

kyle-cypress replied:
Sorry for the delay responding here. At a first pass I think I agree with you that this should in fact support zero for both. I will look at this more closely and confirm once I finish my investigation on #11531 .

kyle-cypress replied:
Pushed bdc1a341e2b07806b32aec85e282ed07a0792de7 to remove this restriction.

Commit:
Expand error checks in _calculate_blocksize_match_tdbstore

The minimum size required by tdbstore is either 2 sectors or 10 pages,
whichever is larger. Correspondingly, adjust the error checks in
_calculate_blocksize_match_tdbstore to match this requirement.

Commit:
TDBStore: Fix potential alignment issue in default addresses

When 10 pages is larger than 2 sectors, align the selected size
down to be an even multiple of the sector size, to ensure that
the allocated space divides cleanly in half for garbage collection.

Commit:
TDBStore: remove get_flash_bounds input constraint

Handle the case where the entirety of flash (size = 0) is required
for a flash memory starting at address 0, instead of erroring out.

Commit:
TDBStore: Increase min pages to 14

Increase minimum page count from 10 to 14 based on further experiments
with features-storage-tests-kvstore-static_tests.

SeppoTakalo commented on features/storage/kvstore/conf/filesystem/mbed_lib.json:
```diff
@@ -36,7 +36,7 @@
     "target_overrides": {
         "MCU_PSOC6": {
-            "rbp_internal_size": 7168
+            "rbp_internal_size": "7168"
```
Why value changed to string?

VeijoPesonen replied:
The default value seems to be also string.

kyle-cypress replied:
Yes, the change to string is for consistency with the type of the default.

SeppoTakalo commented on features/storage/kvstore/conf/filesystem/mbed_lib.json:
```diff
@@ -2,11 +2,11 @@
     "config": {
         "rbp_internal_size": {
-            "help": "Default is the size of the 2 last sectors of internal flash",
+            "help": "Default is the larger of the last 2 sectors or last 10 pages of flash.",
```
You later specify `TDBStore ::STORE_PAGES` to `14`, so why this is 10?

kyle-cypress replied:
Because STORE_PAGES was increased and the comment update was missed. I will fix.

SeppoTakalo commented on features/storage/kvstore/conf/kv_config.cpp:
```diff
@@ -176,13 +176,19 @@ int _calculate_blocksize_match_tdbstore(BlockDevice *bd)
     }
 
+    if (number_of_page < TDBStore::STORE_PAGES) {
+        tr_warning("KV Config: There are less than fourteen pages - TDBStore will not work.");
```
Should be `tr_error()`, and preferably change the previous as well.

Also not sure whether `TDBStore::STORE_PAGES` can change in the future, so perhaps use `less than %d pages`.

SeppoTakalo commented on features/storage/kvstore/conf/kv_config.cpp:
```diff
@@ -266,181 +272,23 @@ FileSystem *_get_filesystem_default(const char *mount)
 }
 
-BlockDevice *_get_blockdevice_SPIF(bd_addr_t start_address, bd_size_t size)
+BlockDevice *_get_blockdevice_SPIF(bd_addr_t &start_address, bd_size_t &size)
```
If we change the API, should we use pointers instead of references?
Makes it more obvious for caller that the values can change. I don't really like references for trivial types.

kyle-cypress replied:
I looked through the current changeset and nothing uses the returned value anymore. I will revert the argument type change.

Commit:
Remove unnecessary reference-typed arguments

No callers make use of the modified argument values, so change
them to a more straightforward pass by value.

Commit:
KV_CONFIG: Change errors to use tr_error not tr_warning

Commit:
Reuse TDBStore default size computation in devicekey test

Replace custom caluation that always assumed two sectors with the standard
calculation exposed on TDBStore.

VeijoPesonen reviewed (approved):
Looks good to me

0xc0170 reviewed (dismissed):
Please add release notes

SeppoTakalo commented on features/storage/kvstore/tdbstore/TDBStore.h:
```diff
@@ -280,6 +287,32 @@ class TDBStore : public KVStore {
+     *          MBED_ERROR_INITIALIZATION_FAILED    Failed to initialize flash driver.
+     *          MBED_ERROR_MEDIA_FULL               Default TDBStore space overlaps with program memory.
+     */
+    static int get_default_flash_addresses(bd_addr_t *start_address, bd_size_t *size);
```
I don't think this is the right place for these new APIs.
`TDBStore` uses just a `BlockDevice`, it does not have to know directly that the block device is going to be `FlashIAP`, and therefore, it does not have to know the flash map either. If we want to keep the class clean, we should not add that dependency here.

Maybe `kv_config.cpp` or `SystemStorage.cpp` would be better place for these.

kyle-cypress replied:
I'll look into moving this. If it doesn't wind up introducing other dependency issues I think kv_config makes more sense because there are some kvstore-specific assumptions in the defaults that this uses, and systemstorage seems to be more generic.

kyle-cypress replied:
Moved into kv_config, rebased on latest master, and pushed. 
With these changes, `mbed test -t ARMC6 -m LPC55S69_S --app-config tools/psa/tfm/mbed_app.json -n *psa-spm_smoke -DUSE_PSA_TEST_PARTITIONS -DUSE_SMOKE_TESTS_PART1 --compile` passes. Thanks.

Commit:
Move flash bounds helpers from TDBStore to kv_config

Commit:
kv_config: Remove hard fail on too few pages

STORE_SECTORS is a hard requirement. If there are fewer than 2 pages
then the kvstore will not work, because the garbage collection process
relies on having at least two sectors to work with.
STORE_PAGES is a heuristic. It is a reasonable default to use if the
application does not specify the amount of flash to use for TDBStore.
But if an application knows that a smaller number of pages will suffice
for its specific needs, then that is valid and should be permitted.

Commit:
Remove stray include of TDBStore.h

Commit:
Add kvstore/conf to unittest includes

The unittests compile DirectAccessDeviceKey.cpp which depends on
kv_config.h, which lives in features/storage/kvstore/conf
